include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,1336 @@
+        m_var_map(std::move(var_map)), m_rev_var_map(std::move(rev_var_map)),
+        m_cache_lines(std::move(cache_lines)), m_regs(std::move(regs)),
+        m_addrs(std::move(addrs)), m_flags(std::move(flags)) {}
+
+  // Perform *this = join(*this, right)
+  void do_join(const mru_region_domain_t &right) {
+    base_varname_allocator_t out_alloc(m_alloc, right.m_alloc);
+    base_uf_domain_t out_addrs(m_addrs | right.m_addrs);
+    base_bool_domain_t out_flags(m_flags | right.m_flags);
+
+    base_abstract_domain_t right_cache_lines(right.m_cache_lines);
+    base_uf_domain_t right_regs(right.m_regs);
+    var_map_t out_var_map;
+    rev_var_map_t out_rev_var_map;
+    // -- Compute common renamings
+    base_variable_vector_t left_vars, right_vars, out_vars;
+    // reserve vec space by upper bound to avoid reallocations
+    size_t num_renamings = m_var_map.size();
+    left_vars.reserve(num_renamings);
+    right_vars.reserve(num_renamings);
+    out_vars.reserve(num_renamings);
+    // perform the common renamings
+    for (auto &kv : m_var_map) {
+      const variable_t &v = kv.first;
+      auto it = right.m_var_map.find(v);
+      if (it != right.m_var_map.end()) {
+        // if we ignore unknown regions, types are the same by construction
+        if (!crab_domain_params_man::get().region_skip_unknown_regions() &&
+            !kv.second.same_type(it->second)) {
+          crab::CrabStats::count(
+              domain_name() + ".count.join.skipped.inconsistent_dynamic_type");
+          continue;
+        }
+        ghost_variables_t out_gvars(
+            ghost_variables_t::create(out_alloc, kv.second));
+        // Store all ghost variables into each vector
+        kv.second.add(left_vars);
+        it->second.add(right_vars);
+        out_gvars.add(out_vars);
+        out_var_map.insert({v, out_gvars});
+        out_gvars.update_rev_varmap(out_rev_var_map, v);
+      }
+    }
+
+    // might need project the variables used in common because common variables
+    // are subset of the variables used in domain
+    m_cache_lines.project(left_vars);
+    m_cache_lines.rename(left_vars, out_vars);
+    right_cache_lines.project(right_vars);
+    right_cache_lines.rename(right_vars, out_vars);
+
+    m_regs.project(left_vars);
+    m_regs.rename(left_vars, out_vars);
+    right_regs.project(right_vars);
+    right_regs.rename(right_vars, out_vars);
+
+    m_cache_lines |= right_cache_lines;
+    m_regs |= right_regs;
+    m_is_bottom = m_cache_lines.is_bottom();
+    std::swap(m_alloc, out_alloc);
+    std::swap(m_var_map, out_var_map);
+    std::swap(m_rev_var_map, out_rev_var_map);
+    std::swap(m_addrs, out_addrs);
+    std::swap(m_flags, out_flags);
+
+    CRAB_LOG("mru-region", crab::outs() << *this << "\n");
+  }
+
+  mru_region_domain_t do_join_or_widening(
+      const mru_region_domain_t &left, const mru_region_domain_t &right,
+      const bool is_join /*unused*/, base_dom_binop_t base_dom_op,
```

Comment:
Be consistent. Either pass a function that performs join or widening on each subdomain or use the bool `is_join`. Right now, you pass the function only for `m_cache_lines` and `m_regs` but not for `m_addrs` , etc. My personal taste is to use `is_join` because it reduces the number of parameters so it's less verbose.

Reply:
this is what I want to use overloading. I saw we pass a function ptr as an argument here. So I expect we can use the template feature to pass just one function. That is, passing a template function (lambda function) as an argument here to allow each individual subdomain to call their corresponded methods. However, I think the template for lambda function is C++ 20 features. I also cannot pass a template function as an argument here. Any idea to achieve this?